src/java/org/apache/cassandra/cql3/restrictions/SingleColumnRestriction.java

```diff
@@ -403,49 +381,80 @@ public void addToRowFilter(RowFilter filter, IndexRegistry indexRegistry, QueryO
             for (Bound b : Bound.values())
                 if (hasBound(b))
                     filter.add(columnDef, slice.getIndexOperator(b), slice.bound(b).bindAndGet(options));
+
+            for (MarkerOrTerms markerOrTerms : skippedValues)
+            {
+                for (ByteBuffer value : markerOrTerms.bindAndGet(options, columnDef.name))
+                   filter.add(columnDef, Operator.NEQ, value);
+            }
         }
 
         @Override
         protected boolean isSupportedBy(Index index)
         {
-            return slice.isSupportedBy(columnDef, index);
+            boolean supportsSlice = slice.isSupportedBy(columnDef, index);
+            boolean supportsNeq = index.supportsExpression(columnDef, Operator.NEQ);
+            return  supportsSlice || !skippedValues.isEmpty() && supportsNeq;
```

Comment:
The implementations of `SingleColumnRestriction.SliceRestriction#isSupportedBy` and `MultiColumnRestriction.SliceRestriction#isSupportedBy` are identical. Also, both implementations are the only callers of `TermSlice#isSupportedBy(ColumnMetadata, Index)`. 

Maybe we could pass the list of skipped values to `TermSlice#isSupportedBy` and delegate the NEQ check to it, putting similar things together and preventing some duplication:
```java
public boolean isSupportedBy(ColumnMetadata column, Index index, List<MarkerOrTerms> skippedValues)
{
    boolean supported = false;

    if (hasBound(Bound.START))
        supported |= isInclusive(Bound.START) ? index.supportsExpression(column, Operator.GTE)
                : index.supportsExpression(column, Operator.GT);
    if (hasBound(Bound.END))
        supported |= isInclusive(Bound.END) ? index.supportsExpression(column, Operator.LTE)
                : index.supportsExpression(column, Operator.LT);

    supported |= !skippedValues.isEmpty() && index.supportsExpression(column, Operator.NEQ);

    return supported;
}
```

Reply:
I don't agree on this one. This is just coincidental duplication - accidentally both single and multi restrictions work the same way here, but there is no fundamental reason they have to. The amount of duplication is also very tiny. And introducing an unrelated parameter to `TermSlice` is mixing responsibilities which is IMHO worse.

I thought that maybe we could move the skipped values to the `TermSlice` itself. Then it would make more sense to have `TermSlice` handle the `isSupportedBy` logic as it would have all the information it needs. WDYT?

Reply:
Agree, moving the skipped values to the `TermSlice` sounds good to me.